### PR TITLE
refactor: New Icon system with Enhanced Antd Custom Icon

### PR DESCRIPTION
**Pull request by geido**
### SUMMARY
This PR introduces an entirely new Icon system that uses a combination of the Antd Custom Icon and Emotion while consolidating the usage of icons.

- Use existing Antd icons with the ability to enhance them
- Use custom icons trough the Antd Custom Icon component and the ability to enhance them
- Adds the ability to override any Antd existing icon with a custom one

As an example implementation, the following sections have been changed to use some of the icons under the new icon system:

- Charts lists
- Dashboards lists
- Saved Queries lists
- Datasets lists
- Databases lists
- Explore - Edit Datasource columns

## BEFORE

charts-list-BEFORE
dashboard-list-trash-BEFORE
dashboard-lists-cardview-BEFORE
databases-lists-BEFORE
datasets-list-BEFORE
saved-query-list-trash-BEFORE
Screen Shot 2021-02-03 at 17 14 41

## AFTER

charts-list-AFTER
dashboard-list-trash-AFTER
dashboard-lists-cardview-AFTER
databases-lists-AFTER
datasets-list-AFTER
saved-query-list-trash-AFTER
Screen Shot 2021-02-03 at 20 56 09

### TEST PLAN
1. Go to any list
2. Hover your mouse on a list item
3. Make sure the trash icon appears and works as intended

### ADDITIONAL INFORMATION

- [ ] Has associated issue:
- [x] Changes UI
- [ ] Requires DB Migration.
- [ ] Confirm DB Migration upgrade and downgrade tested.
- [x] Introduces new feature or API
- [ ] Removes existing feature or API

### CODE REVIEW

**Comment by geido** [comment]
@rusackas @junlincc

**Comment by adam-stasiak** [comment]
Tested manually across browsers and looks fine 🟢

**Comment by rusackas** [comment]
For transparency, Diego and I spoke online, and we agreed to put this in Draft mode while exploring a couple options:

• Re-exporting vanilla AntD icons from the Superset Icon component, so we can get ALL icons from one import.
• Attempting to consolidate the interface for components, either with AntD's `<WhateverIcon />` interface, or our own `<Icon name="WhateverIcon" />` interface.
• Once the interface is conformed, trying to find a way to essentially override an AntD icon, which would allow a component developer to use `AnyVanillaIcon` from AntD, and then when we want to replace it with our own, we would simply _override_ it in our `Icon` component, and not have to go around finding/replacing the scattered implementations.

File: superset-frontend/src/components/Icons/index.ts [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+import AntdEnhancedIcons from './all/AntdEnhanced';
+import Trash from './all/Trash';
+import Alert from './all/Alert';
+import AlertSolid from './all/AlertSolid';
+import AlertSolidSmall from './all/AlertSolidSmall';
+import Binoculars from './all/Binoculars';
+import Bolt from './all/Bolt';
+import BoltSmall from './all/BoltSmall';
+import BoltSmallRun from './all/BoltSmallRun';
+
```
Comment (rusackas) [inline superset-frontend/src/components/Icons/index.ts:29]: I think this PR is looking AMAZING, but there's one thing I'm trying to weigh (and I'm curious what you think). 

Rather than having a whole bunch of small icon files, would it be easier to grow/maintain the icon set if we just made this file longer, doing all the VG imports and `<Icon ...>` composition in one place? ¯\\\_(ツ)_/¯ 

I think what you have done here is very clean and proper... I'm just wondering if the other way would make life easier. Kind of a nitpick, but now's the time to think about it I suppose.

File: superset-frontend/src/views/CRUD/chart/ChartCard.tsx [hunk 2]
```diff
@@ -102,22 +102,26 @@ export default function ChartCard({
...
       )}
       {canExport && (
-        <Menu.Item
-          role="button"
-          tabIndex={0}
-          onClick={() => handleBulkChartExport([chart])}
-        >
-          <ListViewCard.MenuIcon name="share" /> {t('Export')}
+        <Menu.Item>
+          <div
```
Comment (rusackas) [inline superset-frontend/src/views/CRUD/chart/ChartCard.tsx:106]: Curious why this change is needed, only in case it's an indication of things we'll have to do elsewhere.

File: superset-frontend/src/views/CRUD/dashboard/DashboardList.tsx [hunk 3]
```diff
@@ -301,10 +302,7 @@ function DashboardList(props: DashboardListProps) {
                         className="action-button"
                         onClick={confirmDelete}
                       >
-                        <Icon
-                          data-test="dashboard-list-trash-icon"
-                          name="trash"
-                        />
+                        <Icons.Trash data-test="dashboard-list-trash-icon" />
```
Comment (rusackas) [inline superset-frontend/src/views/CRUD/dashboard/DashboardList.tsx:305]: I really like this implementation, btw. Seems much cleaner/clearer.

File: superset-frontend/src/components/ListView/ActionsBar.tsx [hunk 4]
```diff
@@ -53,6 +54,7 @@ export default function ActionsBar({ actions }: ActionsBarProps) {
   return (
     <StyledActions className="actions">
       {actions.map((action, index) => {
+        const ActionIcon = Icons[action.icon];
```
Comment (rusackas) [inline superset-frontend/src/components/ListView/ActionsBar.tsx:57]: nifty!

File: superset-frontend/src/components/Icons/index.ts (hunk 1 above)
Comment (nytai) [reply superset-frontend/src/components/Icons/index.ts:29]: I second this, I'd prefer not to touch 3 files (svg, component file, and this file) when adding a new svg icon

File: superset-frontend/src/views/CRUD/chart/ChartCard.tsx (hunk 2 above)
Comment (geido) [reply superset-frontend/src/views/CRUD/chart/ChartCard.tsx:106]: @rusackas Antd adds a span around the SVG. There might be cases where related styles affect the span causing the Icon to behave incorrectly. This was one of those cases. As a container DIV was used elsewhere in this file, I went ahead and made it consistent everywhere. By doing so the style override issue gets resolved as well.

File: superset-frontend/src/components/Icons/index.ts (hunk 1 above)
Comment (geido) [reply superset-frontend/src/components/Icons/index.ts:150]: @rusackas @nytai I tend to agree as well. I have made some changes and now the index imports, enhances, and exports everything back. This makes the file a bit lengthy but it keeps all the functionalities discussed plus what you want here. This is a work-in-progress and I still need to finish re-exporting all the existing icons, but I'd appreciate your feedback while I move forward with this.

Comment (nytai) [reply superset-frontend/src/components/Icons/index.ts:150]: This is looking pretty good. It may get messy eventually, but that might just mean we need to clean up our icons. I think it's pretty clear what changes are needed to add a new icon from glancing at this file.

**Review by rusackas — APPROVED** [review]
LGTM! This is a huge step forward! Will pull the branch and do a little more functional testing again before merging

**Comment by rusackas** [comment]
This looks like it's ready to go, pending a teensy `lint-fix`

**Comment by rusackas** [comment]
@junlincc @eschutho @nytai 

With concerns of stability in mind, does anyone see any reason to further hesitate on merging this one in (after a rebase)? It looks like a very positive code quality change to me, so I'm getting an itchy trigger finger.

File: superset-frontend/src/components/Icons/index.ts [hunk 5]
```diff
@@ -0,0 +1,284 @@
...
+import { ReactComponent as ListIcon } from 'images/icons/list.svg';
+import { ReactComponent as ListViewIcon } from 'images/icons/list_view.svg';
+import { ReactComponent as LocationIcon } from 'images/icons/location.svg';
+import { ReactComponent as LockLockedIcon } from 'images/icons/lock_locked.svg';
+import { ReactComponent as LockUnlockedIcon } from 'images/icons/lock_unlocked.svg';
+import { ReactComponent as MapIcon } from 'images/icons/map.svg';
+import { ReactComponent as MessageIcon } from 'images/icons/message.svg';
+import { ReactComponent as MinusIcon } from 'images/icons/minus.svg';
+import { ReactComponent as MinusSolidIcon } from 'images/icons/minus_solid.svg';
+import { ReactComponent as ModeHorizIcon } from 'images/icons/more_horiz.svg';
```
Comment (villebro) [inline superset-frontend/src/components/Icons/index.ts:106]: nit: this should probably be `MoreHorizIcon`?

**Review by villebro — APPROVED** [review]
One minor nit - other than that LGTM and worked perfectly in my testing.

**Comment by eschutho** [comment]
cc @steejay @mihir174 
Looks great. Only question I have is whether we can we still use custom icons when designers want something new or different?

**Comment by rusackas** [comment]
> cc @Steejay @mihir174
> Looks great. Only question I have is whether we can we still use custom icons when designers want something new or different?

Indeed! This is built specifically for that, so we can effectively extend/override the AntD icon set, rather than having custom icons that are a bit of an interface mismatch from the AntD icons we'd use elsewhere.
